### PR TITLE
Rework XStream initialization

### PR DESCRIPTION
**Pull request by xerus2000**
Due to the tests now running in parallel, we sometimes had ConcurrentModificationExceptions since the global XStream instance was initialized in parallel from different threads.

Now all initialization points are locked.

### CODE REVIEW

**Review by anarchuser — COMMENTED** [review]
If the problem is concurrent invocation of `registerXStream`, wouldn't it be more sensible to make `GamePlugin` a singleton?

**Comment by xerus2000** [comment]
That function is already static. That doesn't solve the concurrency problem, unfortunately.

**Comment by xerus2000** [comment]
I have run all tests repeatedly locally with no failure. The one failing test has been unreliable for a while, I'll see whether I have to Disable it or find a fix.

**Comment by anarchuser** [comment]
As far as I can tell, that function is only called by the ctor of GamePlugin, right?

**Comment by xerus2000** [comment]
you're right, the concurrency issues occur only in the tests. Hmm...

File: plugin/src/server/sc/plugin2021/GamePlugin.kt [hunk 1]
```diff
@@ -31,9 +31,14 @@ class GamePlugin: IGamePlugin {
                     Color::class.java, Team::class.java)
         
         fun registerXStream() {
-            if (loaded.compareAndSet(false, true)) {
+            if (loaded.compareAndSet(-1, 0)) {
```
Comment (SKoschnicke) [inline plugin/src/server/sc/plugin2021/GamePlugin.kt:34]: Please add a comment explaining that this is required for thread safety.

**Review by SKoschnicke — APPROVED** [review]
Tests are succeeding on my machine with the current state :+1:

File: plugin/src/server/sc/plugin2021/GamePlugin.kt (hunk 1 above)
Comment (xerus2000) [reply plugin/src/server/sc/plugin2021/GamePlugin.kt:34]: Yes, I am a bit unsure now actually. This issue is _only_ relevant for the parallelized tests, in normal usage there is no thread safety needed. So this seems a bit over the top...

**Comment by anarchuser** [comment]
I think the actual issue is still why that function is being called several types at all

**Comment by xerus2000** [comment]
As said, parallelized tests

**Comment by anarchuser** [comment]
yes, but this should be setup stage for tests. This shouldn't be something several tests initialise for themselves anyways.

**Comment by xerus2000** [comment]
yeah, the question is whether there is a way to have global setup for independent tests.

**Comment by xerus2000** [comment]
hello it's me, giant commit :smile: yes I just changed a lot of code. I will also add some documentation for that in the GUIDELINES soon. But unlike last time, where I made the XStream instance global (brevity at the expense of creating global state, which is the opposite of clean code ^^), this really does make the code a whole lot cleaner (which we can partly see through the application of the design patterns Factory + Template Method)!

**Review by SKoschnicke — APPROVED** [review]
Awesome!

**Pull request merged**